Userland/Libraries/LibGUI/DynamicWidgetContainer.cpp

```diff
@@ -0,0 +1,464 @@
+            }
+
+            for (auto const& child : new_child_order)
+                add_child(*child);
+        } else {
+            int persisted_state = Config::read_i32(config_domain(), "DynamicWidgetContainers"sv, section_label(), (int)ViewState::Expanded);
+            set_view_state(static_cast<ViewState>(persisted_state));
+        }
+        update();
+    });
+}
+
+void DynamicWidgetContainer::set_section_label(String label)
+{
+    m_section_label = move(label);
+
+    auto label_widget = m_controls_widget->find_descendant_of_type_named<GUI::LabelWithEventDispatcher>("section_label");
+    label_widget->set_text(m_section_label);
+}
+
+void DynamicWidgetContainer::set_config_domain(String domain)
+{
+    m_config_domain = move(domain);
+    // FIXME: A much better solution would be to call the restore_view_state within a dedicated "initialization finished" method that is called by the gml runtime after that widget is ready.
+    //        We do not have such a method yet.
+    restore_view_state();
+}
+
+void DynamicWidgetContainer::set_detached_size(Gfx::IntSize const size)
+{
+    m_detached_size = { size };
+}
+
+void DynamicWidgetContainer::set_show_controls(bool value)
+{
+    m_show_controls = value;
+    m_controls_widget->set_visible(m_controls_widget->is_visible() && show_controls());
+    update();
+}
+
+void DynamicWidgetContainer::set_container_with_individual_order(bool value)
+{
+    m_is_container_with_individual_order = value;
+}
+
+void DynamicWidgetContainer::second_paint_event(PaintEvent&)
+{
+    GUI::Painter painter(*this);
+    painter.fill_rect_with_gradient(Gfx::IntRect(0, height() - 1, width(), height()), palette().threed_shadow1().lightened(), palette().button());
+
+    if (!m_is_dragging && !m_render_as_move_target)
+        return;
+
+    if (m_is_dragging) {
+        // FIXME: Would be nice if we could paint outside our own boundaries.
+        auto move_widget_indicator = rect().translated(m_current_mouse_position).translated(-m_drag_start_location);
+        painter.fill_rect(move_widget_indicator, palette().rubber_band_fill());
+        painter.draw_rect_with_thickness(move_widget_indicator, palette().rubber_band_border(), 1);
+    } else if (m_render_as_move_target) {
+        painter.fill_rect(rect(), palette().rubber_band_fill());
+        painter.draw_rect_with_thickness({ rect().x(), rect().y(), rect().width() - 1, rect().height() - 1 }, palette().rubber_band_border(), 1);
+    }
+}
+
+ErrorOr<void> DynamicWidgetContainer::detach_widgets()
+{
+    if (!m_detached_widgets_window.has_value()) {
+        auto detached_window = TRY(GUI::Window::try_create());
+        detached_window->set_title(section_label().to_deprecated_string());
+        detached_window->set_window_type(WindowType::Normal);
+        if (m_detached_size.has_value())
+            detached_window->resize(detached_size());
+        else
+            detached_window->resize(size());
```

Comment:
Mixing and matching `m_detached_size` and `detached_size()` feels awkward. Maybe `detached_size()` should just return an Optional size always?

Reply:
That is a bit clunky because it is used as a size property and therfore does not support the optional :/ I've introduced a `has_detached_size()` function.